channelacceptor/rpcacceptor.go

```diff
@@ -0,0 +1,180 @@
+type ChanAcceptorClient struct {
+	SendChan chan *OpenChannelRequest
+	RecvChan chan OpenChannelResponse
+
+	closeChan chan struct{}
+	active    bool
+}
+
+// RPCAcceptor represents the RPC-controlled variant of the ChannelAcceptor.
+// Only one RPC client is currently allowed at a time.
+type RPCAcceptor struct {
+	clientMtx sync.Mutex
+	client    *ChanAcceptorClient
+
+	acceptChan    chan *OpenChannelRequest
+	acceptClients map[[32]byte]chan bool
+
+	quit chan struct{}
+	wg   sync.WaitGroup
+}
+
+// Accept is a predicate on the OpenChannelRequest which is sent to the RPC
+// client who will respond with the ultimate decision. This assumes a registered
+// client exists.
+func (r *RPCAcceptor) Accept(req *OpenChannelRequest) bool {
+	// Register a new channel for the PendingChannelID associated with this
+	// OpenChannelRequest.
+	responseChan := make(chan bool)
+	r.acceptClients[req.OpenChanMsg.PendingChannelID] = responseChan
+
+	select {
+	case r.acceptChan <- req:
+	case <-r.quit:
+		return false
+	}
+
+	for {
+		select {
+		case accept, _ := <-responseChan:
+			return accept
+		case <-time.After(DefaultRPCTimeout):
+			return false
+		case <-r.quit:
+			return false
+		}
+	}
+}
+
+// RegisterClient registers an RPC streaming client. There may only be one RPC
+// streaming client at a time.
+func (r *RPCAcceptor) RegisterClient() (*ChanAcceptorClient, error) {
+	r.clientMtx.Lock()
+	defer r.clientMtx.Unlock()
+
+	if r.active() {
+		return nil, fmt.Errorf("client already exists")
+	}
+
+	r.client = &ChanAcceptorClient{
+		SendChan:  make(chan *OpenChannelRequest),
+		RecvChan:  make(chan OpenChannelResponse),
+		closeChan: make(chan struct{}),
+		active:    true,
+	}
+
+	return r.client, nil
+}
+
+// UnregisterClient unregisters the RPC streaming client.
+func (r *RPCAcceptor) UnregisterClient() {
+	r.clientMtx.Lock()
+	defer r.clientMtx.Unlock()
+
+	close(r.client.closeChan)
+	r.client.active = false
+
+	return
+}
+
+// Start starts the RPCAcceptor and allows the RPC server to register clients.
+func (r *RPCAcceptor) Start() error {
+	r.acceptChan = make(chan *OpenChannelRequest)
+	r.acceptClients = make(map[[32]byte]chan bool)
+	r.quit = make(chan struct{})
+
+	r.wg.Add(1)
+	go r.processRequests()
+
+	return nil
+}
+
+// Stop stops the RPCAcceptor.
+func (r *RPCAcceptor) Stop() error {
+	close(r.quit)
+	r.wg.Wait()
+
+	return nil
+}
+
+// processRequests receives OpenChannelRequests from the Accept method and
+// forwards them to the chan between the RPC server. It also facilitates
+// forwarding the OpenChannelResponses from the RPC server back to the funding
+// manager.
+func (r *RPCAcceptor) processRequests() {
+	defer r.wg.Done()
+	for {
+		select {
+		case <-r.quit:
+			return
+		default:
+		}
+
+		// If a client does not exist, continue.
+		if !r.active() {
+			continue
+		}
+
+		select {
+		case rpcResp := <-r.client.RecvChan:
+			// This is an OpenChannelResponse received from the RPC server. We
```

Comment:
Would lift the comment here and below to reside right on top of the top-level `select` case.